### PR TITLE
Github API support

### PR DESCRIPTION
**Pull request by mcuadros**
This PR allows to navpatch retrieve the content of the repository via Github API.

### CODE REVIEW

File: navpatch/repository.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
+package navpatch
+
+import "strings"
+
+type Repository interface {
+	GetTree() (TreeEntry, error)
```
Comment (tcard) [inline navpatch/repository.go]: http://golang.org/doc/effective_go.html#Getters

> [...] the getter method should be called Owner (upper case, exported), not GetOwner.

File: navpatch/navigator.go [hunk 2]
```diff
@@ -14,25 +14,25 @@ import (
...
 	if err != nil {
-		return nil, fmt.Errorf("walking directory %s: %s", baseDir, err)
+		return nil, err
 	}
 
 	changes := ApplyChangesToTree(patchSet, tree)
 
 	return &Navigator{
-		BasePath: baseDir,
+		BasePath: "foo",
```
Comment (tcard) [inline navpatch/navigator.go]: Leftover.

File: navpatch/navigator.go [hunk 3]
```diff
@@ -14,25 +14,25 @@ import (
...
 	patchSet, err := patch.Parse(rawPatch)
 	if err != nil {
 		return nil, fmt.Errorf("parsing patch: %s", err)
 	}
 
-	tree, err := DirPathToTree(baseDir)
+	tree, err := r.GetTree()
 	if err != nil {
-		return nil, fmt.Errorf("walking directory %s: %s", baseDir, err)
+		return nil, err
```
Comment (tcard) [inline navpatch/navigator.go:28]: You should probably add a method to `Repository` that gives you the base path or URL.

Please leave the `fmt.Errorf` as it was. Adding context to the error returned from a deeper level is clarifying and done across the whole codebase.

File: navpatch/repositories/fs.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
+package repositories
```
Comment (tcard) [inline navpatch/repositories/fs.go:1]: I'd rather keep a flat package structure here unless there's a good reason not to.

Comment (mcuadros) [inline navpatch/repositories/fs.go:1]: the goal is keep isolated the several implementation of the same thing in a package. And avoid file names as `repository_fs.go`

File: main.go [hunk 5]
```diff
@@ -29,6 +36,18 @@ func main() {
...
+func buildRepository(path string) (navpatch.Repository, error) {
+	if _, err := os.Stat(path); err == nil {
+		return repositories.NewFSRepository(path), nil
+	}
+
+	if _, err := vcsurl.Parse(path); err == nil {
+		return repositories.NewGithubRepository(path)
+	}
+
+	return nil, fmt.Errorf("Invalid path or VCS url: %s ", path)
```
Comment (tcard) [inline main.go]: It's conventional for errors to start with lowercase. Also, trailing space.

File: main.go [hunk 6]
```diff
@@ -8,14 +8,21 @@ import (
...
 func main() {
 	listenAddr, baseDir, rawPatch := processArgs()
 
-	nav, err := navpatch.NewNavigator(baseDir, rawPatch)
+	r, err := buildRepository(baseDir)
+	if err != nil {
+		internal.ErrorExit(err)
+	}
+
+	nav, err := navpatch.NewNavigator(r, rawPatch)
```
Comment (tcard) [inline main.go]: `buildRepository` should be done inside `NewNavigator`. The command should be a very thin wrapper for the library. Users of navpatch as a library also want that convenience. This kind of "stringly-typed" interface is commonplace in the standard library.

**Comment by tcard** [comment]
Thanks!

**Pull request merged**